CMakeLists.txt

```diff
@@ -0,0 +1,164 @@
+        src/exceptions/i3_ipc_unsupported.cpp
+        src/i3_json_parser/i3_json_parser-events.cpp
+        src/i3_json_parser/i3_json_parser-requests.cpp
+        src/i3_json_parser/i3_json_parser-utility.cpp
+        src/i3_message.cpp
+)
+target_compile_features(i3-ipc++-obj PUBLIC cxx_std_17)
+target_compile_options(i3-ipc++-obj
+    PRIVATE
+        -O2 -Wall -Wextra -Wshadow -Wold-style-cast -Wcast-align -Wunused -Wpedantic -Wconversion
+        -Wsign-conversion -Wdouble-promotion -Wformat=2 -Werror -Wfatal-errors
+)
+target_link_libraries(i3-ipc++-obj PRIVATE "${RAPIDJSON_CXX_FLAGS}")
+target_include_directories(i3-ipc++-obj
+    PUBLIC
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include/>
+        $<INSTALL_INTERFACE:include/>
+    PRIVATE
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/src/>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/src/i3_json_parser/>
+        $<INSTALL_INTERFACE:src/>
+        $<INSTALL_INTERFACE:src/i3_json_parser/>
+        ${RAPIDJSON_INCLUDE_DIRS}
+)
+set_target_properties(i3-ipc++-obj PROPERTIES POSITION_INDEPENDENT_CODE 1)
+
+option(build_dynamic "Build using shared libraries" ON)
+if (build_dynamic)
+    add_library(i3-ipc++-shared SHARED $<TARGET_OBJECTS:i3-ipc++-obj>)
+    set_target_properties(i3-ipc++-shared
+        PROPERTIES
+            POSITION_INDEPENDENT_CODE 1
+            OUTPUT_NAME i3-ipc++
+    )
+    target_include_directories(i3-ipc++-shared
+        PUBLIC
+            $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include/>
+            $<INSTALL_INTERFACE:include/>
+    )
+    list(APPEND build_targets_list i3-ipc++-shared)
+    if (NOT CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
+        add_library(i3-ipc++::i3-ipc++-shared ALIAS i3-ipc++-shared)
+    endif()
+endif()
+option(build_static "Build using static libraries" ON)
+if (build_static)
+    add_library(i3-ipc++-static STATIC $<TARGET_OBJECTS:i3-ipc++-obj> ${headers})
+    set_target_properties(i3-ipc++-static PROPERTIES OUTPUT_NAME i3-ipc++)
+    target_include_directories(i3-ipc++-static
+        PUBLIC
+            $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include/>
+            $<INSTALL_INTERFACE:include/>
+    )
+    list(APPEND build_targets_list i3-ipc++-static)
+    if (NOT CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
+        add_library(i3-ipc++::i3-ipc++-static ALIAS i3-ipc++-static)
+    endif()
+endif()
+
+add_library(i3-ipc++ INTERFACE)
+target_link_libraries(i3-ipc++ INTERFACE ${build_targets_list})
+if (NOT CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
+    add_library(i3-ipc++::i3-ipc++ ALIAS i3-ipc++)
+endif()
+
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#                            build examples                            #
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+option(
+    build_examples
+    "Build example programs which use this library."
+    TRUE
+)
+if (build_examples)
+    set(EXAMPLES
```

Comment:
I would move this entire block into a separate CMakeLists.txt within examples folder and just do a:
```
if (build_examples)
    add_subdirectory(examples)
endif()
```